cdap-watchdog/src/main/java/co/cask/cdap/logging/read/DistributedLogReader.java

```diff
@@ -17,41 +17,88 @@
 package co.cask.cdap.logging.read;
 
 import co.cask.cdap.common.logging.LoggingContext;
+import co.cask.cdap.logging.appender.kafka.KafkaTopic;
+import co.cask.cdap.logging.appender.kafka.StringPartitioner;
 import co.cask.cdap.logging.filter.Filter;
+import co.cask.cdap.logging.save.CheckpointManager;
+import co.cask.cdap.logging.save.CheckpointManagerFactory;
+import co.cask.cdap.logging.save.KafkaLogWriterPlugin;
 import com.google.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
- * Reads logs in a distributed setup.
+ * Reads logs in a distributed setup, using kafka for latest logs and files for older logs.
  */
 public final class DistributedLogReader implements LogReader {
+  private static final Logger LOG = LoggerFactory.getLogger(DistributedLogReader.class);
+
   private final KafkaLogReader kafkaLogReader;
   private final FileLogReader fileLogReader;
+  private final CheckpointManager checkpointManager;
+  private final StringPartitioner partitioner;
 
   /**
    * Creates a DistributedLogReader object.
    */
   @Inject
-  public DistributedLogReader(KafkaLogReader kafkaLogReader, FileLogReader fileLogReader) {
+  public DistributedLogReader(KafkaLogReader kafkaLogReader, FileLogReader fileLogReader,
+                              CheckpointManagerFactory checkpointManagerFactory, StringPartitioner partitioner) {
     this.kafkaLogReader = kafkaLogReader;
     this.fileLogReader = fileLogReader;
+    this.checkpointManager = checkpointManagerFactory.create(KafkaTopic.getTopic(),
+                                                             KafkaLogWriterPlugin.CHECKPOINT_ROW_KEY_PREFIX);
+    this.partitioner = partitioner;
   }
 
   @Override
   public void getLogNext(final LoggingContext loggingContext, final LogOffset fromOffset, final int maxEvents,
                               final Filter filter, final Callback callback) {
+    // If latest logs are not requested, try reading from file.
+    if (fromOffset != LogOffset.LATEST_OFFSET) {
```

Comment:
One problem with this logic is that we might not full fill the maxEvents requested. Maybe open a JIRA to fix this in 3.1.0